File: backend/app/api/v2/mta.py
"""
Material Transfer Agreement (MTA) API

Endpoints for managing MTAs for germplasm exchange.
"""

from fastapi import APIRouter, HTTPException, Query
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

from app.modules.seed_bank.mta import mta_service, MTAType, MTAStatus, BenefitSharingType

router = APIRouter(prefix="/mta", tags=["Material Transfer Agreements"])


# Pydantic Models
class InstitutionInfo(BaseModel):
    institution: str
    country: str
    contact: str
    email: str


class BenefitSharingInfo(BaseModel):
    type: BenefitSharingType
    details: str
    royalty_rate: Optional[float] = None
    milestones: Optional[list] = None


class CreateMTARequest(BaseModel):
    mta_type: MTAType
    provider: InstitutionInfo
    recipient: InstitutionInfo
    accessions: list[str]
    crops: list[str]
    purpose: str
    benefit_sharing: Optional[BenefitSharingInfo] = None
    exchange_id: Optional[str] = None


class ApproveRequest(BaseModel):
    approver: str


class RejectRequest(BaseModel):
    reason: str
    rejector: str


class SignRequest(BaseModel):
    signatory: str


class TerminateRequest(BaseModel):
    reason: str
    terminator: str


# Endpoints
@router.get("/templates")
async def get_mta_templates():
    """Get all available MTA templates"""
    return {
        "templates": mta_service.get_templates(),
        "types": [t.value for t in MTAType],
    }


@router.get("/templates/{template_id}")
async def get_mta_template(template_id: str):
    """Get specific MTA template with full clauses"""
    template = mta_service.get_template(template_id)
    if not template:
        raise HTTPException(status_code=404, detail="Template not found")
    return template


@router.post("")
async def create_mta(request: CreateMTARequest):
    """Create a new Material Transfer Agreement"""
    try:
        mta = mta_service.create_mta(
            mta_type=request.mta_type,
            provider=request.provider.model_dump(),
            recipient=request.recipient.model_dump(),
            accessions=request.accessions,
            crops=request.crops,
            purpose=request.purpose,
            benefit_sharing=request.benefit_sharing.model_dump() if request.benefit_sharing else None,
            exchange_id=request.exchange_id,
        )
        return {"message": "MTA created successfully", "mta": mta}
    except Exception as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("")
async def list_mtas(
    mta_type: Optional[MTAType] = Query(None, description="Filter by MTA type"),
    status: Optional[MTAStatus] = Query(None, description="Filter by status"),
    institution: Optional[str] = Query(None, description="Filter by institution name"),
):
    """List all MTAs with optional filters"""
    mtas = mta_service.list_mtas(
        mta_type=mta_type,
        status=status,
        institution=institution,
    )
    return {"mtas": mtas, "count": len(mtas)}


@router.get("/statistics")
async def get_mta_statistics():
    """Get MTA statistics"""
    return mta_service.get_statistics()


@router.get("/by-number/{mta_number}")
async def get_mta_by_number(mta_number: str):
    """Get MTA by its number"""
    mta = mta_service.get_mta_by_number(mta_number)
    if not mta:
        raise HTTPException(status_code=404, detail="MTA not found")
    return mta


@router.get("/by-exchange/{exchange_id}")
async def get_mta_for_exchange(exchange_id: str):
    """Get MTA associated with a germplasm exchange"""
    mta = mta_service.get_mta_for_exchange(exchange_id)
    if not mta:
        raise HTTPException(status_code=404, detail="No MTA found for this exchange")
    return mta


@router.get("/{mta_id}")
async def get_mta(mta_id: str):
    """Get MTA by ID"""
    mta = mta_service.get_mta(mta_id)
    if not mta:
        raise HTTPException(status_code=404, detail="MTA not found")
    return mta


@router.post("/{mta_id}/submit")
async def submit_mta_for_review(mta_id: str):
    """Submit MTA for review"""
    try:
        mta = mta_service.submit_for_review(mta_id)
        if not mta:
            raise HTTPException(status_code=404, detail="MTA not found")
        return {"message": "MTA submitted for review", "mta": mta}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{mta_id}/approve")
async def approve_mta(mta_id: str, request: ApproveRequest):
    """Approve MTA after review"""
    try:
        mta = mta_service.approve_mta(mta_id, request.approver)
        if not mta:
            raise HTTPException(status_code=404, detail="MTA not found")
        return {"message": "MTA approved", "mta": mta}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{mta_id}/reject")
async def reject_mta(mta_id: str, request: RejectRequest):
    """Reject MTA"""
    try:
        mta = mta_service.reject_mta(mta_id, request.reason, request.rejector)
        if not mta:
            raise HTTPException(status_code=404, detail="MTA not found")
        return {"message": "MTA rejected", "mta": mta}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{mta_id}/sign")
async def sign_mta(mta_id: str, request: SignRequest):
    """Sign and activate MTA"""
    try:
        mta = mta_service.sign_mta(mta_id, request.signatory)
        if not mta:
            raise HTTPException(status_code=404, detail="MTA not found")
        return {"message": "MTA signed and activated", "mta": mta}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.post("/{mta_id}/terminate")
async def terminate_mta(mta_id: str, request: TerminateRequest):
    """Terminate an active MTA"""
    try:
        mta = mta_service.terminate_mta(mta_id, request.reason, request.terminator)
        if not mta:
            raise HTTPException(status_code=404, detail="MTA not found")
        return {"message": "MTA terminated", "mta": mta}
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))


@router.get("/{mta_id}/compliance")
async def check_mta_compliance(mta_id: str):
    """Check MTA compliance status"""
    result = mta_service.check_compliance(mta_id)
    if "error" in result:
        raise HTTPException(status_code=404, detail=result["error"])
    return result


@router.get("/types/reference")
async def get_mta_types():
    """Get MTA type reference data"""
    return {
        "types": [
            {
                "value": MTAType.SMTA.value,
                "name": "Standard MTA (SMTA)",
                "description": "For ITPGRFA Annex I crops, auto-approved",
            },
            {
                "value": MTAType.INSTITUTIONAL.value,
                "name": "Institutional MTA",
                "description": "Bilateral agreement between institutions",
            },
            {
                "value": MTAType.RESEARCH.value,
                "name": "Research MTA",
                "description": "Research-only use, no commercial development",
            },
            {
                "value": MTAType.COMMERCIAL.value,
                "name": "Commercial MTA",
                "description": "Commercial development with benefit sharing",
            },
        ],
        "statuses": [
            {"value": s.value, "name": s.value.replace("_", " ").title()}
            for s in MTAStatus
        ],
        "benefit_sharing_types": [
            {"value": b.value, "name": b.value.replace("_", " ").title()}
            for b in BenefitSharingType
        ],
    }

File: frontend/src/lib/api/seed-bank/mta.ts
import { ApiClientCore } from "../core/client";

export interface MTAInstitutionInfo {
  institution: string;
  country: string;
  contact: string;
  email: string;
}

export interface MTABenefitSharingInfo {
  type: "none" | "monetary" | "non_monetary" | "both";
  details: string;
  royalty_rate?: number;
  milestones?: Array<Record<string, unknown>>;
}

export interface MTARecord {
  id: string;
  mta_number: string;
  type: "smta" | "institutional" | "research" | "commercial";
  status: "draft" | "pending_review" | "pending_signature" | "active" | "expired" | "terminated" | "rejected";
  provider: MTAInstitutionInfo;
  recipient: MTAInstitutionInfo;
  accessions: string[];
  accession_count: number;
  crops: string[];
  purpose: string;
  benefit_sharing: MTABenefitSharingInfo;
  created_date: string;
  signed_date?: string | null;
  effective_date?: string | null;
  expiry_date?: string | null;
  exchange_id?: string | null;
}

export interface MTATemplate {
  id: string;
  name: string;
  type: MTARecord["type"];
  description: string;
  version?: string;
  clause_count: number;
  benefit_sharing: MTABenefitSharingInfo["type"];
  duration_years: number | null;
  requires_approval: boolean;
}

export interface MTAStatistics {
  total: number;
  active: number;
  pending: number;
  institutions: number;
  by_type: Record<string, number>;
  by_status: Record<string, number>;
  total_accessions_under_mta: number;
}

export interface MTAComplianceResult {
  mta_id: string;
  mta_number: string;
  compliant: boolean;
  issues: string[];
  warnings: string[];
  checked_at: string;
}

export interface MTAListResponse {
  mtas: MTARecord[];
  count: number;
}

export interface MTATemplatesResponse {
  templates: MTATemplate[];
  types: string[];
}

export interface CreateMTAPayload {
  mta_type: MTARecord["type"];
  provider: MTAInstitutionInfo;
  recipient: MTAInstitutionInfo;
  accessions: string[];
  crops: string[];
  purpose: string;
  benefit_sharing?: MTABenefitSharingInfo;
  exchange_id?: string | null;
}

interface MTAActionResponse {
  message: string;
  mta: MTARecord;
}

export class MTAService {
  constructor(private client: ApiClientCore) {}

  async getMTAs(params: { mtaType?: string; status?: string; institution?: string } = {}) {
    const searchParams = new URLSearchParams();
    if (params.mtaType) searchParams.append("mta_type", params.mtaType);
    if (params.status) searchParams.append("status", params.status);
    if (params.institution) searchParams.append("institution", params.institution);

    const query = searchParams.toString();
    return this.client.get<MTAListResponse>(`/api/v2/mta${query ? `?${query}` : ""}`);
  }

  async getStatistics() {
    return this.client.get<MTAStatistics>("/api/v2/mta/statistics");
  }

  async getTemplates() {
    return this.client.get<MTATemplatesResponse>("/api/v2/mta/templates");
  }

  async createMTA(payload: CreateMTAPayload) {
    return this.client.post<MTAActionResponse>("/api/v2/mta", payload);
  }

  async submitMTA(mtaId: string) {
    return this.client.post<MTAActionResponse>(`/api/v2/mta/${mtaId}/submit`);
  }

  async approveMTA(mtaId: string, approver: string) {
    return this.client.post<MTAActionResponse>(`/api/v2/mta/${mtaId}/approve`, { approver });
  }

  async rejectMTA(mtaId: string, reason: string, rejector: string) {
    return this.client.post<MTAActionResponse>(`/api/v2/mta/${mtaId}/reject`, { reason, rejector });
  }

  async signMTA(mtaId: string, signatory: string) {
    return this.client.post<MTAActionResponse>(`/api/v2/mta/${mtaId}/sign`, { signatory });
  }

  async terminateMTA(mtaId: string, reason: string, terminator: string) {
    return this.client.post<MTAActionResponse>(`/api/v2/mta/${mtaId}/terminate`, { reason, terminator });
  }

  async getCompliance(mtaId: string) {
    return this.client.get<MTAComplianceResult>(`/api/v2/mta/${mtaId}/compliance`);
  }
}